### PR TITLE
[Go Bootcamp] First deliverable

### PR DESCRIPTION
**Pull request by jvillarreal-w**
Had some repetition concerns towards error handling, other than that I just followed the steps of the clean architecture article and added some complementary info to the CSV, such as base stats, so I can fetch that same data from PokeAPI for the next deliverable and not have it be just names.

### CODE REVIEW

File: common/constants.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package common
+
+const (
+	InternalDataSourcePath = "pkmn.csv"
```
Comment (StradivariT) [inline common/constants.go:4]: Keeping this kind of info separated is fine, although it's a lot more commonly found to have it on a config file, and then when the app initializes you would like to read the file with this and potentially more values, like the external api endpoints, db tables, etc...

File: infrastructure/router/router.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+	"github.com/jvillarreal-w/academy-go-q42021/interface/controller"
+	"github.com/labstack/echo"
+	"github.com/labstack/echo/middleware"
+)
+
+func NewRouter(e *echo.Echo, c controller.AppController) *echo.Echo {
+	e.Use(middleware.Logger())
+	e.Use(middleware.Recover())
+
+	e.GET("/pokemon", func(context echo.Context) error { return c.Pokemon.GetPokemon(context) })
```
Comment (StradivariT) [inline infrastructure/router/router.go:13]: I'm not familiar with echo, but would it be possible to just have, for example: 
`e.GET("/pokemon", c.Pokemon.GetPokemon(context))` ?

Just to avoid redundancy

Comment (StradivariT) [reply infrastructure/router/router.go:13]: Same for the route below

File: interface/controller/app_controller.go [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package controller
+
+type AppController struct {
+	Pokemon interface{ PokemonController }
```
Comment (StradivariT) [inline interface/controller/app_controller.go]: Are you planning on adding more interfaces to the Pokemon attribute? If not you could just remove the `interface` wrapper

File: usecase/interactor/pkmn_interactor.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	Get(p []*model.Pokemon) ([]*model.Pokemon, error)
+	GetById(p *model.Pokemon, id string) (*model.Pokemon, error)
+}
+
+func NewPokemonInteractor(r repository.PokemonRepository) PokemonInteractor {
+	return &pokemonInteractor{r}
+}
+
+func (ps *pokemonInteractor) Get(p []*model.Pokemon) ([]*model.Pokemon, error) {
+	p, err := ps.PokemonRepository.FindAll(p)
```
Comment (StradivariT) [inline usecase/interactor/pkmn_interactor.go]: If you don't need to do any more logic after this call then it's ok to just `return ps.PokemonRepository.FindAll(p)`.

Same for the function below

File: interface/repository/pkmn_repository.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+	fmt.Println("Successfully opened CSV file")
+	defer f.Close()
+
+	rows, err := csv.NewReader(f).ReadAll()
+	if err != nil {
+		fmt.Println(err)
```
Comment (StradivariT) [inline interface/repository/pkmn_repository.go]: I think you would also like to return here

File: interface/controller/pkmn_controller.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package controller
```
Comment (StradivariT) [inline interface/controller/pkmn_controller.go:1]: This comment applies for all the errors in general.

While it's ok to return when an error happens, what it's usually done is to log the error you got and then respond to the user with an error that would be more appropriate.

Logging allows us to debug from our side what's going on since logs are usually aggregated into a logging service, and it also allows us to not return an error so explicit to the user that could potentially give out details on how our APIs work. So after logging whatever error we got, we want to return to the user a custom error.

For example, if you didn't find the resource the user was requesting it would make sense to return a 404 - not found with a message that goes like `pokemon not found`, or when something internal failed, like failing to read the csv file, or failing to connect with a DB, then we would just send the user a 500 status code with a message like `something went wrong!`, while logging as much info as we can to help us debug if necessary.

So TL;DR, log your errors and return something more user friendly as a response

File: interface/repository/pkmn_repository.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/jvillarreal-w/academy-go-q42021/common"
+	"github.com/jvillarreal-w/academy-go-q42021/domain/model"
+	"github.com/jvillarreal-w/academy-go-q42021/usecase/repository"
+)
+
+type pokemonRepository struct {
```
Comment (StradivariT) [inline interface/repository/pkmn_repository.go:15]: I would highly suggest having the file path as an attribute here on your struct, that way when initializing a new Repository, you could provide any file path, which will make unit testing a lot easier since your not binding all of this logic to just one file. We don't want to make tests with the actual real file, but with test files, testing will come later but I thought it would be worth mentioning this now

File: interface/controller/pkmn_controller.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+	if err != nil {
+		return err
+	}
+
+	return c.JSON(http.StatusOK, p)
+}
+
+func (pc *pokemonController) GetPokemonById(c Context) error {
+	var p *model.Pokemon
+	id := c.Param("id")
```
Comment (StradivariT) [inline interface/controller/pkmn_controller.go:39]: One of the main responsibilites of a controller is to validate if the data you receive from the user is valid before moving on, if you need to convert this to an int later on, the place to do it is here on your controller, and not in deeper layers of your app. So if the id is not valid, you could return an error response to the user indicating that the id provided is not valid

File: interface/repository/pkmn_repository.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+		return nil, err
+	}
+
+	pkmnId, err := strconv.ParseUint(id, 10, 32)
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, row := range rows {
```
Comment (StradivariT) [inline interface/repository/pkmn_repository.go]: Seems like there's a whole bunch of repeated logic in here, could you try to move the repeated stuff in a reusable function? 

One suggestion I would make is that, if you already have a function that retrieves all of the pokemons, then in this function that finds one by its ID you could call that function and then just search for the requested ID here.

**Review by StradivariT — CHANGES_REQUESTED** [review]
Overall looks great, it also seems like it's on a good track to continue with the 2nd deliverable stuff with little changes.

There's stil a couple of things to be done to be able to do unit testing on an ideal way, but so far excellent job!

File: infrastructure/router/router.go (hunk 2 above)
Comment (jvillarreal-w) [reply infrastructure/router/router.go:13]: Seems like the intended way of declaring routes with Echo, though to be fair I cannot say I am familiar with this either.
https://echo.labstack.com/guide/routing/

File: interface/controller/app_controller.go (hunk 3 above)
Comment (jvillarreal-w) [reply interface/controller/app_controller.go]: Addressed in a more recent commit, thank you.

File: usecase/interactor/pkmn_interactor.go (hunk 4 above)
Comment (jvillarreal-w) [reply usecase/interactor/pkmn_interactor.go]: Addressed in a more recent commit, thank you.

File: interface/repository/pkmn_repository.go (hunk 5 above)
Comment (jvillarreal-w) [reply interface/repository/pkmn_repository.go]: Addressed as well, thank you.

File: interface/controller/pkmn_controller.go (hunk 6 above)
Comment (jvillarreal-w) [reply interface/controller/pkmn_controller.go:1]: Added a general purpose logger to address this.

File: interface/repository/pkmn_repository.go (hunk 7 above)
Comment (jvillarreal-w) [reply interface/repository/pkmn_repository.go:15]: Added in a later commit, thank you.

File: interface/controller/pkmn_controller.go (hunk 8 above)
Comment (jvillarreal-w) [reply interface/controller/pkmn_controller.go:39]: Now validated in the controller.

File: interface/repository/pkmn_repository.go (hunk 9 above)
Comment (jvillarreal-w) [reply interface/repository/pkmn_repository.go]: Thanks for the suggestion, that seems to do it.
